packages/@aws-cdk/assertions/lib/private/resources.ts

```diff
@@ -17,9 +18,9 @@ export function findResources(inspector: StackInspector, type: string, props: an
   return result.matches;
 }
 
-export function hasResource(inspector: StackInspector, type: string, props: any): string | void {
+export function hasResource(inspector: StackInspector, type: string, props: any, addProperties?: boolean): string | void {
   const section: { [key: string]: Resource } = inspector.value.Resources;
-  const result = matchSection(filterType(section, type), props);
```

Comment:
Can we instead slice this out into a method called `hasResourceProperties()` that does the necessary coercion and calls `hasResource()`? Feels like that would be cleaner than shimmying a boolean here.

Reply:
I sorta disagree that that would be cleaner (but will do it if you double down on your opinion). I agree that this doesn't look very elegant. However, the alternative is having two functions with very similar bodies. I don't think its feasible for `hasResourceProperties()` to call `hasResource()` after it has added the empty Property objects. What would have to happen is a third internal function, say `matchResources()` which both `hasResourceProperties()` and `hasResource()` would call. And this complicates the special case of `hasResourceProperties('type', Match.absent())` that we agreed to support, since this would have to avoid adding empty property objects. It would look like this --

```ts
// private/resource.ts
hasResource(inspector: StackInspector, type: string, props: any): string | void {
  const section: { [key: string]: Resource } = inspector.value.Resources;
  return matchSection(filterType(section, type), type, props);
}

hasResourceProperties(inspector: StackInspector, type: string, props: any): string | void {
  const section: { [key: string]: Resource } = inspector.value.Resources;
  return matchSection(addEmptyProperties(filterType(section, type)), type, props);
}

matchSection(section: any, type: string, props: any): string | void {
  if (result.match) {
    return;
  }

  if (result.closestResult === undefined) {
    return `No resource with type ${type} found`;
  }

  return [
    `Template has ${result.analyzedCount} resources with type ${type}, but none match as expected.`,
    formatFailure(result.closestResult),
  ].join('\n');
}
```

And then of course the exposed API for `hasResourceProperties()` would have to account for the special case:

```ts
// template.ts
  public hasResourceProperties(type: string, props: any): void {
    if (props instanceof AbsentMatch) {
        return hasResource(this.inspector, type, Match.absent());
    }
    const matchError = hasResourceProperties(this.inspector, type, Match.objectLike({
      Properties: Matcher.isMatcher(props) ? props : Match.objectLike(props),
    }));
    if (matchError) {
      throw new Error(matchError);
    }
  }
```

So -- I think that the current implementation saves a lot of code and is easier to follow than the alternative. And since this is an internal API we do not have to worry about discoverability and all that jazz.

Laying out the options I can see. Happy to go with whichever you prefer.

Reply:
I'm having a hard time following what the problem is here.

Why not something like this -

```ts
hasResourceProperties(template: Template, type: string, props: any): string | void {
  const amended = addEmptyProperties(template);
  return hasResource(amended, type, props);
}
```

Reply:
Good call. I think this was actually impossible with the previous version of the API with `inspector` as a parameter rather than `template`. It would be problematic to modify `inspector.template` since the modification would be persist beyond the particular test. Here we can deep copy `template`, modify it, and then send it to `hasResource`.